src/main/scala/net/ripe/rpki/publicationserver/store/XodusObjectStore.scala

```diff
@@ -0,0 +1,100 @@
+package net.ripe.rpki.publicationserver.store
+
+import java.net.URI
+
+import com.softwaremill.macwire.MacwireMacros._
+import jetbrains.exodus.entitystore.{Entity, StoreTransaction, StoreTransactionalComputable, StoreTransactionalExecutable}
+import net.ripe.rpki.publicationserver._
+import net.ripe.rpki.publicationserver.model.ClientId
+
+import scala.collection.JavaConversions._
+import scala.concurrent.Future
+
+class XodusObjectStore extends Hashing {
+
+  lazy val conf: AppConfig = wire[AppConfig]
+
+  import XodusDB._
+
+  private val OBJECT_ENTITY_NAME = "object"
+
+  type RRDPObject = (Base64, Hash, URI, ClientId)
+
+  private def inTx(f: StoreTransaction => Unit): Unit = {
+    entityStore.executeInTransaction(new StoreTransactionalExecutable() {
+      override def execute(txn: StoreTransaction): Unit = f(txn)
+    })
+  }
+
+  private def withReadTx[T](f: StoreTransaction => T): T = {
+    entityStore.computeInReadonlyTransaction(new StoreTransactionalComputable[T]() {
+      override def compute(txn: StoreTransaction): T = f(txn)
+    })
+  }
+
+  private type StoredTuple = (String, String, String, String)
+
+  private def insertAction(txn: StoreTransaction, obj: RRDPObject): Unit = {
+    val (base64, hash, uri, clientId) = obj
+    val e = txn.newEntity("object")
+    fillEntity(base64, hash, uri, clientId, e)
+  }
+
+  private def fillEntity(base64: Base64, hash: Hash, uri: URI, clientId: ClientId, e: Entity) = {
+    e.setProperty("base64", base64.value)
+    e.setProperty("hash", hash.hash)
+    e.setProperty("uri", uri.toString)
+    e.setProperty("clientId", clientId.value)
+  }
+
+  private def updateAction(txn: StoreTransaction, obj: RRDPObject) = {
+    val (base64, hash, uri, clientId) = obj
+    txn.find(OBJECT_ENTITY_NAME, "uri", uri.toString).
+      foreach(e => fillEntity(base64, hash, uri, clientId, e))
+  }
+
+  private def deleteAction(txn: StoreTransaction, hash: Hash) = {
+    txn.find(OBJECT_ENTITY_NAME, "hash", hash.hash).
+      foreach(e => e.delete())
+  }
+
+  def clear() = inTx { txn =>
+    txn.getAll(OBJECT_ENTITY_NAME).foreach(e => e.delete())
+  }
+
+  def getState: XodusObjectStore.State = withReadTx { txn =>
+    txn.getAll(OBJECT_ENTITY_NAME).map { e =>
+      val base64 = Base64(e.getProperty("base64").toString)
+      val hash = Hash(e.getProperty("hash").toString)
+      val uri = URI.create(e.getProperty("uri").toString)
+      val clientId = ClientId(e.getProperty("clientId").toString)
+      uri -> (base64, hash, clientId)
+    }.toMap
+  }
+
+  def applyChanges(changeSet: QueryMessage, clientId: ClientId): Future[Unit] =
+    Future.successful {
```

Comment:
My Scala is rusty by now, but why is this Future immediately successful (and not just Future?)
```suggestion
    Future {
```

Reply:
Good point. I removed the Future completely, as it's only requested by Slick and we don't really use the async nature of the API.